rust/parser/test/mod.rs

```diff
@@ -427,7 +416,144 @@ $x isa movie,
     has tmdb-vote-count <= 400;"#;
 
     let parsed = parse_query(query).unwrap().into_match();
-    let expected = typeql_match!(var_concept("x").isa("movie").has(("tmdb-vote-count", lte(400))));
+    let expected = typeql_match!(cvar("x").isa("movie").has(("tmdb-vote-count", lte(400))));
+
+    assert_valid_eq_repr!(expected, parsed, query);
+}
+
+#[test]
+fn test_attribute_query_by_value_variable() {
+    let query = r#"match
+?x = 5;
+$a == ?x isa age;"#;
+
+    let parsed = parse_query(query).unwrap().into_match();
+    let expected = typeql_match!(vvar("x").assign(Expression::Constant(5.into())), cvar("a").eq(vvar("x")).isa("age"),);
+
+    assert_valid_eq_repr!(expected, parsed, query);
+}
+
+#[test]
+fn test_variable_name_clash_throws() {
+    let query = r#"match\n
+$z isa person, has age $y;
+?y = $y;"#;
+
+    let parsed = parse_query(query);
+    assert!(parsed.is_err());
+}
+
+#[test]
+fn test_assign_ops() {
+    let query = r#"match
+$x isa commodity,
+    has price $p;
+(commodity: $x, qty: $q) isa order;
+?net = $p * $q;
+?gross = ?net * ( 1.0 + 0.21 );"#;
+
+    let parsed = parse_query(query).unwrap().into_match();
+    let expected = typeql_match!(
+        cvar("x").isa("commodity").has(("price", cvar("p"))),
+        rel(("commodity", "x")).rel(("qty", "q")).isa("order"),
+        vvar("net").assign(cvar("p").multiply(cvar("q"))),
+        vvar("gross").assign(vvar("net").multiply(1.0.add(0.21))),
```

Comment:
I've got mixed feelings about `1.0.add(0.21)`
One one hand, this is neat. On the other, this kind of clashes with `f64 as Add<f64>` [1]. I'm not prepared to say that this is a problem: if the user writes `1.add(2)` in the builder and TypeQL produces either `3` or `1+2`, those should be indistinguishable to the user. I'm not sure the law of least surprise kicks in at any point here, but that remains to be seen for smth like strings.

[1] <https://doc.rust-lang.org/stable/std/primitive.f64.html#impl-Add%3Cf64%3E-for-f64>

Reply:
We can create a function `constant` like `cvar` and `vvar` and use `constant(1).add(2)` instead of `1.add(2)`.